File: cli/src/core/references/sources/LinearAdapter.ts
/**
 * LinearAdapter — first concrete `SourceAdapter`.
 *
 * Wraps the historical Linear extraction algorithm (see
 * `cli/src/core/references/ReferenceExtractor.ts`) behind the adapter interface so the
 * extractor main loop can drive every source uniformly. Behaviour is
 * byte-identical to `formatLinearIssuesBlock` / `tryBuildRef` so existing
 * Linear tests pass unchanged.
 */

import type { Reference, ReferenceField } from "../../../Types.js";
import { escapeForAttr, escapeForText } from "../../PromptXmlEscape.js";
import type { SourceAdapter } from "./SourceAdapter.js";

const LINEAR_TICKET_ID_REGEX = /^[A-Z][A-Z0-9_]*-\d+$/;
const URL_REGEX = /^https?:\/\//;
const DEFAULT_MAX_CHARS = 4000;
const DEFAULT_MAX_TOTAL = 30000;

function isObject(v: unknown): v is Record<string, unknown> {
	return typeof v === "object" && v !== null && !Array.isArray(v);
}

function readPriority(obj: Record<string, unknown>): string | undefined {
	const p = obj.priority;
	if (typeof p === "string" && p.length > 0) return p;
	if (isObject(p)) {
		const name = (p as { name?: unknown }).name;
		if (typeof name === "string" && name.length > 0) return name;
	}
	return undefined;
}

function readLabels(obj: Record<string, unknown>): ReadonlyArray<string> | undefined {
	const l = obj.labels;
	if (!Array.isArray(l)) return undefined;
	const strs = l.filter((x): x is string => typeof x === "string" && x.length > 0);
	return strs.length > 0 ? strs : undefined;
}

function truncate(s: string, max: number): string {
	if (s.length <= max) return s;
	return `${s.slice(0, max)}\n…[truncated, ${s.length - max} more chars]`;
}

/** Build the Linear-specific display fields. Source knowledge lives here only. */
function buildFields(
	status: string | undefined,
	priority: string | undefined,
	labels: ReadonlyArray<string> | undefined,
): ReferenceField[] {
	const fields: ReferenceField[] = [];
	if (status !== undefined)
		fields.push({ key: "status", label: "Status", value: status, icon: "circle-large-filled" });
	if (priority !== undefined) fields.push({ key: "priority", label: "Priority", value: priority, icon: "flame" });
	if (labels !== undefined && labels.length > 0) {
		fields.push({ key: "labels", label: "Labels", value: labels.join(", "), icon: "tag" });
	}
	return fields;
}

export const LinearAdapter: SourceAdapter = {
	id: "linear",
	mcpPrefix: "mcp__linear__",
	wrapperKeys: ["items", "issues", "nodes", "results"],
	maxCharsPerReference: DEFAULT_MAX_CHARS,

	extractRef(payload, toolName, referencedAt) {
		if (!isObject(payload)) return null;
		const obj = payload as Record<string, unknown>;
		const id = obj.id;
		const title = obj.title;
		const url = obj.url;
		if (typeof id !== "string" || !LINEAR_TICKET_ID_REGEX.test(id)) return null;
		if (typeof title !== "string" || title.length === 0) return null;
		if (typeof url !== "string" || !URL_REGEX.test(url)) return null;

		const status = typeof obj.status === "string" && obj.status.length > 0 ? obj.status : undefined;
		const priority = readPriority(obj);
		const labels = readLabels(obj);
		const description =
			typeof obj.description === "string" && obj.description.length > 0 ? obj.description : undefined;

		const fields = buildFields(status, priority, labels);
		return {
			mapKey: `linear:${id}`,
			source: "linear",
			nativeId: id,
			title,
			url,
			...(fields.length > 0 ? { fields } : {}),
			...(description !== undefined ? { description } : {}),
			toolName,
			referencedAt,
		};
	},

	renderPromptBlock(refs, opts) {
		if (refs.length === 0) return "";
		const maxPer = opts?.maxCharsPerReference ?? DEFAULT_MAX_CHARS;
		const maxTotal = opts?.maxTotalChars ?? DEFAULT_MAX_TOTAL;
		const sorted = [...refs].sort((a, b) => a.referencedAt.localeCompare(b.referencedAt));
		const reversed = [...sorted].reverse();
		const selected: Reference[] = [];
		let total = 0;
		for (const r of reversed) {
			const rendered = renderOne(r, maxPer);
			if (total + rendered.length > maxTotal) break;
			selected.push(r);
			total += rendered.length;
		}
		if (selected.length === 0) return "";
		selected.reverse();
		return `<linear-issues>\n${selected.map((r) => renderOne(r, maxPer)).join("\n")}\n</linear-issues>`;
	},
};

function renderOne(ref: Reference, maxChars: number): string {
	const attrs = [`id="${escapeForAttr(ref.nativeId)}"`];
	for (const f of ref.fields ?? []) attrs.push(`${f.key}="${escapeForAttr(f.value)}"`);
	const lines = [`<issue ${attrs.join(" ")}>`];
	lines.push(`  <title>${escapeForText(ref.title)}</title>`);
	lines.push(`  <url>${escapeForText(ref.url)}</url>`);
	if (ref.description) {
		lines.push("  <description>");
		lines.push(escapeForText(truncate(ref.description, maxChars)));
		lines.push("  </description>");
	}
	lines.push("</issue>");
	return lines.join("\n");
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/references/LinearAdapter.kt
package ai.jolli.jollimemory.core.references

import com.google.gson.JsonObject

/**
 * LinearAdapter — [SourceAdapter] for the Linear MCP server.
 *
 * Validates `id` matches `^[A-Z][A-Z0-9_]*-\d+$`, reads
 * status/priority/labels/description from the payload.
 */
object LinearAdapter : SourceAdapter {

	private val TICKET_ID_RE = Regex("^[A-Z][A-Z0-9_]*-\\d+$")
	private val URL_RE = Regex("^https?://")
	private const val DEFAULT_MAX_CHARS = 4000
	private const val DEFAULT_MAX_TOTAL = 30000

	override val id = SourceId.linear
	override val maxCharsPerReference = DEFAULT_MAX_CHARS
	override val wrapperKeys = listOf("items", "issues", "nodes", "results")

	override fun extractRef(payload: JsonObject, toolName: String, referencedAt: String): Reference? {
		val ticketId = payload.stringOrNull("id") ?: return null
		if (!TICKET_ID_RE.matches(ticketId)) return null
		val title = payload.stringOrNull("title")
		if (title.isNullOrEmpty()) return null
		val url = payload.stringOrNull("url") ?: return null
		if (!URL_RE.containsMatchIn(url)) return null

		val status = payload.stringOrNull("status")?.ifEmpty { null }
		val priority = readPriority(payload)
		val labels = readLabels(payload)
		val description = payload.stringOrNull("description")?.ifEmpty { null }

		val fields = buildFields(status, priority, labels)
		return Reference(
			mapKey = "linear:$ticketId",
			source = SourceId.linear,
			nativeId = ticketId,
			title = title,
			url = url,
			fields = fields.ifEmpty { null },
			description = description,
			toolName = toolName,
			referencedAt = referencedAt,
		)
	}

	override fun renderPromptBlock(refs: List<Reference>, opts: RenderOptions): String {
		if (refs.isEmpty()) return ""
		val maxPer = opts.maxCharsPerReference ?: DEFAULT_MAX_CHARS
		val maxTotal = opts.maxTotalChars ?: DEFAULT_MAX_TOTAL
		val selected = selectByBudget(refs, maxPer, maxTotal) { renderOne(it, maxPer) }
		if (selected.isEmpty()) return ""
		return "<linear-issues>\n${selected.joinToString("\n") { renderOne(it, maxPer) }}\n</linear-issues>"
	}

	private fun readPriority(obj: JsonObject): String? {
		val p = obj.get("priority") ?: return null
		if (p.isJsonPrimitive && p.asJsonPrimitive.isString) {
			val s = p.asString
			return s.ifEmpty { null }
		}
		if (p.isJsonObject) {
			val name = p.asJsonObject.stringOrNull("name")
			return name?.ifEmpty { null }
		}
		return null
	}

	private fun readLabels(obj: JsonObject): List<String>? {
		val arr = obj.arrayOrNull("labels") ?: return null
		val strs = arr.filter { it.isJsonPrimitive && it.asJsonPrimitive.isString }
			.map { it.asString }
			.filter { it.isNotEmpty() }
		return strs.ifEmpty { null }
	}

	private fun buildFields(
		status: String?,
		priority: String?,
		labels: List<String>?,
	): List<ReferenceField> {
		val fields = mutableListOf<ReferenceField>()
		if (status != null) fields.add(ReferenceField("status", "Status", status, "circle-large-filled"))
		if (priority != null) fields.add(ReferenceField("priority", "Priority", priority, "flame"))
		if (!labels.isNullOrEmpty()) fields.add(ReferenceField("labels", "Labels", labels.joinToString(", "), "tag"))
		return fields
	}

	private fun renderOne(ref: Reference, maxChars: Int): String = buildString {
		val attrs = mutableListOf("id=\"${PromptXmlEscape.escapeForAttr(ref.nativeId)}\"")
		ref.fields?.forEach { f -> attrs.add("${f.key}=\"${PromptXmlEscape.escapeForAttr(f.value)}\"") }
		appendLine("<issue ${attrs.joinToString(" ")}>")
		appendLine("  <title>${PromptXmlEscape.escapeForText(ref.title)}</title>")
		appendLine("  <url>${PromptXmlEscape.escapeForText(ref.url)}</url>")
		if (ref.description != null) {
			appendLine("  <description>")
			appendLine(PromptXmlEscape.escapeForText(truncate(ref.description, maxChars)))
			appendLine("  </description>")
		}
		append("</issue>")
	}
}

/** Select refs by char budget, most-recent first, then re-sort chronologically. */
internal fun selectByBudget(
	refs: List<Reference>,
	maxPer: Int,
	maxTotal: Int,
	render: (Reference) -> String,
): List<Reference> {
	val sorted = refs.sortedBy { it.referencedAt }
	val reversed = sorted.reversed()
	val selected = mutableListOf<Reference>()
	var total = 0
	for (r in reversed) {
		val rendered = render(r)
		if (total + rendered.length > maxTotal) break
		selected.add(r)
		total += rendered.length
	}
	selected.reverse()
	return selected
}

internal fun truncate(s: String, max: Int): String {
	if (s.length <= max) return s
	return "${s.substring(0, max)}\n…[truncated, ${s.length - max} more chars]"
}
